### PR TITLE
generate H2O_GIT_REVISION in compile time

### PR DESCRIPTION
**Pull request by i110**
Amends https://github.com/h2o/h2o/pull/1665: define new target `gitrev` which generate `include/h2o/gitrev.h`, and conditionally include it in version.in.

Some notes and considerations:
- ~~If the compiler doesn't have `__has_include` macro, revision number doesn't appended~~
- Users also can directly pass a make argument like `make -DH2O_GIT_REVISION=$(git rev-parse --short HEAD)`
- ~~Which targets should depends on `gitrev` ? (currently only `ADD_DEPENDENCIES(h2o gitrev)`)~~

### CODE REVIEW

File: misc/generate_gitrev.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+GITREV=$(git rev-parse --short HEAD)
+GITTAG=$(git tag --points-at ${GITREV})
```
Comment (kazuho) [inline misc/generate_gitrev.sh]: Do we need error checks (e.g. `git` command returning an error)?

I wouldn't mind if you wrote this script using Perl, if use of sh makes you negative to having error checks.

File: version.h.in [hunk 2]
```diff
@@ -22,6 +22,10 @@
 #ifndef h2o__version_h
 #define h2o__version_h
 
+#ifdef HAS_GITREV_H
```
Comment (kazuho) [inline version.h.in]: Please add `H2O_` prefix.

I would also appreciate it if you could consider using same naming conventions for the names of the file and variables related.
It seems that now the files (generate_gitrev.sh and gitrev.h use "gitrev"), while the variable defined in the file uses "GIT_REVISION". Can't they always be "gitrev" ("GITREV") or "git-revision" ("GIT_REVISION")?

**Comment by i110** [comment]
@kazuho Thank you for your comment. I addressed the issues.

File: misc/generate_gitrev.pl [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env perl
+use strict;
+use warnings;
+
+my $gitrev = exec_git_command('git rev-parse --short HEAD') or die "failed to get current revision";
+my $gittag = exec_git_command("git tag --points-at $gitrev");
+my $content = $gittag ? '' : "#define H2O_GITREV $gitrev";
```
Comment (kazuho) [inline misc/generate_gitrev.pl]: Should we better add `\n` **unless** it is an empty line?

**Comment by i110** [comment]
@kazuho Thank you for your review. I addressed it and merged.

**Comment by skaji** [comment]
After this PR was merged, I could not build h2o. Is this expected?
```
❯ cmake .
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libssl.so;/usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.0.2g")
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.8")
-- Performing Test ARCH_SUPPORTS_64BIT_ATOMICS
-- Performing Test ARCH_SUPPORTS_64BIT_ATOMICS - Success
-- Performing Test LIBC_HAS_BACKTRACE
-- Performing Test LIBC_HAS_BACKTRACE - Success
-- Checking for module 'libuv>=1.0.0'
--
-- Could NOT find LIBUV (missing:  LIBUV_LIBRARIES LIBUV_INCLUDE_DIR)
-- Checking for module 'libwslay'
--   No package 'libwslay' found
-- Could NOT find WSLAY (missing:  WSLAY_LIBRARIES WSLAY_INCLUDE_DIR)
CMake Error at CMakeLists.txt:742 (ADD_DEPENDENCIES):
  Cannot add target-level dependencies to non-existent target
  "t-00unit-libuv.t".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.


-- Configuring incomplete, errors occurred!
See also "/home/skaji/src/github.com/h2o/h2o/CMakeFiles/CMakeOutput.log".
See also "/home/skaji/src/github.com/h2o/h2o/CMakeFiles/CMakeError.log".
```

**Comment by i110** [comment]
@skaji sorry it's my fault. Fixed by https://github.com/h2o/h2o/pull/1689

**Pull request merged**